File: Seminar/src/main/java/org/sopt/Seminar/domain/post/controller/PostController.java
package org.sopt.Seminar.domain.post.controller;

import java.net.URI;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.sopt.Seminar.domain.post.dto.PostCreateRequest;
import org.sopt.Seminar.domain.post.dto.PostGetResponse;
import org.sopt.Seminar.domain.post.dto.PostUpdateRequest;
import org.sopt.Seminar.domain.post.service.PostService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/post")
public class PostController {
    private final PostService postService;
    private static final String CUSTOM_AUTH_ID = "X-Auth-Id";
    private static final String POST_API_ENDPOINT = "/api/post/";

    @PostMapping
    public ResponseEntity<Void> createPost(@RequestHeader(CUSTOM_AUTH_ID) Long memberId,
                                           @RequestBody PostCreateRequest request) {
        URI location = URI.create(POST_API_ENDPOINT + postService.create(request, memberId));
        return ResponseEntity.created(location).build();
    }

    @GetMapping("{postId}")
    public ResponseEntity<PostGetResponse> getPostById(@PathVariable Long postId) {
        return ResponseEntity.ok().body(postService.getById(postId));
    }

    @GetMapping
    public ResponseEntity<List<PostGetResponse>> getPosts(@RequestHeader(CUSTOM_AUTH_ID) Long memberId) {
        return ResponseEntity.ok().body(postService.getPosts(memberId));
    }

    @PatchMapping("{postId}")
    public ResponseEntity<Void> updatePost(@PathVariable Long postId,
                                           @RequestBody PostUpdateRequest request) {
        postService.editContent(postId, request);
        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("{postId}")
    public ResponseEntity<Void> deletePost(@PathVariable Long postId) {
        postService.deleteById(postId);
        return ResponseEntity.noContent().build();
    }
}

File: kopring/week2/kotlin-semina/src/main/kotlin/org/sopt/kotlinsemina/domain/post/controller/PostController.kt
package org.sopt.kotlinsemina.domain.post.controller

import jakarta.validation.Valid
import org.sopt.kotlinsemina.domain.post.dto.PostCreateRequest
import org.sopt.kotlinsemina.domain.post.dto.PostGetResponse
import org.sopt.kotlinsemina.domain.post.dto.PostUpdateRequest
import org.sopt.kotlinsemina.domain.post.service.PostService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.net.URI
import java.security.Principal

@RestController
@RequestMapping("/api/post")
class PostController(
    private val postService: PostService
) {
    @PostMapping
    fun createPost(@Valid @RequestBody request: PostCreateRequest,
                   principal: Principal) : ResponseEntity<Void> {
        val memberId = principal.name.toLong()
        val location = URI.create(POST_API_ENDPOINT + postService.create(request, memberId))

        return ResponseEntity.created(location).build()
    }

    @GetMapping("/{postId}")
    fun getPostById(@PathVariable postId: Long): ResponseEntity<PostGetResponse> {
        return ResponseEntity.ok().body(postService.getById(postId))
    }

    @GetMapping
    fun getPosts(@RequestHeader(CUSTOM_AUTH_ID) memberId: Long): ResponseEntity<List<PostGetResponse>> {
        return ResponseEntity.ok().body(postService.getPosts(memberId))
    }

    @PatchMapping("/{postId}")
    fun updatePost(@PathVariable postId: Long,
                   @RequestBody request: PostUpdateRequest): ResponseEntity<Void> {
        postService.editContent(postId, request)
        return ResponseEntity.noContent().build()
    }

    @DeleteMapping("/{postId}")
    fun deletePost(@PathVariable postId: Long): ResponseEntity<Void> {
        postService.deleteById(postId)
        return ResponseEntity.noContent().build()
    }

    companion object {
        private const val CUSTOM_AUTH_ID: String = "X-Auth-Id"
        private const val POST_API_ENDPOINT: String = "/api/post/"
    }
}